sway/desktop/transaction.c

```diff
@@ -0,0 +1,292 @@
+		}
+	}
+	return false;
+}
+
+void transaction_add_container(struct sway_transaction *transaction,
+		struct sway_container *container) {
+	if (transaction_has_container(transaction, container)) {
+		return;
+	}
+	struct sway_transaction_instruction *instruction =
+		calloc(1, sizeof(struct sway_transaction_instruction));
+	instruction->transaction = transaction;
+	instruction->container = container;
+
+	copy_pending_state(container, &instruction->state);
+
+	if (container->type == C_VIEW) {
+		save_view_buffer(container->sway_view, instruction);
+	}
+	list_add(transaction->instructions, instruction);
+}
+
+void transaction_add_damage(struct sway_transaction *transaction,
+		struct wlr_box *_box) {
+	struct wlr_box *box = calloc(1, sizeof(struct wlr_box));
+	memcpy(box, _box, sizeof(struct wlr_box));
+	list_add(transaction->damage, box);
+}
+
+/**
+ * Apply a transaction to the "current" state of the tree.
+ */
+static void transaction_apply(struct sway_transaction *transaction) {
+	int i;
+	// Apply the instruction state to the container's current state
+	for (i = 0; i < transaction->instructions->length; ++i) {
+		struct sway_transaction_instruction *instruction =
+			transaction->instructions->items[i];
+		struct sway_container *container = instruction->container;
+
+		// There are separate children lists for each instruction state, the
+		// container's current state and the container's pending state
+		// (ie. con->children). The list itself needs to be freed here.
+		// Any child containers which are being deleted will be cleaned up in
+		// transaction_destroy().
+		list_free(container->current.children);
+
+		memcpy(&container->current, &instruction->state,
+				sizeof(struct sway_container_state));
+	}
+
+	// Apply damage
+	for (i = 0; i < transaction->damage->length; ++i) {
+		struct wlr_box *box = transaction->damage->items[i];
+		for (int j = 0; j < root_container.children->length; ++j) {
+			struct sway_container *output = root_container.children->items[j];
+			output_damage_box(output->sway_output, box);
+		}
+	}
+}
+
+static int handle_timeout(void *data) {
+	struct sway_transaction *transaction = data;
+	wlr_log(L_DEBUG, "Transaction %p timed out (%li waiting), applying anyway",
+			transaction, transaction->num_waiting);
+	transaction_apply(transaction);
+	transaction_destroy(transaction);
+	return 0;
+}
+
+void transaction_commit(struct sway_transaction *transaction) {
+	wlr_log(L_DEBUG, "Transaction %p committing with %i instructions",
+			transaction, transaction->instructions->length);
+	transaction->num_waiting = 0;
+	for (int i = 0; i < transaction->instructions->length; ++i) {
+		struct sway_transaction_instruction *instruction =
+			transaction->instructions->items[i];
+		struct sway_container *con = instruction->container;
+		if (con->type == C_VIEW && !con->destroying &&
+				(con->current.view_width != instruction->state.view_width ||
+				 con->current.view_height != instruction->state.view_height)) {
```

Comment:
These parens aren't necessary, || has a higher precedence than &&

Reply:
Removing these parens also makes the code unreadable.

Reply:
nah

Reply:
lol I'm dumb